### ISSUE TITLE
Frozen Core G2 Segfault

### ISSUE THREAD

**Issue by JonathonMisiewicz** [psi4/psi4#1530]
There is a segfault when `set freeze_core true` is set before a `gaussian-2` computation. Note that because of how `gaussian-2` is defined, defining a basis set is not necessary. The segfault occurs after `CGRSolver converged.` prints. I'm not familiar enough with G2 to know whether frozen core G2 is defined, or if this keyword needs to be ignored.

```
molecule {
o
h 1 r1
h 1 r1 2 a1

r1 = 1.0 
a1 = 104.5
}

set freeze_core true

energy('gaussian-2')
```

@edeprince3

**Comment by andysim** [comment]
It [looks like](https://pubs.acs.org/doi/pdf/10.1021/jp9802432) the MP2 part should have the core correlated (even though the basis set is ill suited to describing the core), with the single point energy being performed with a frozen core.  If that's prescribed by the method, setting the `freeze_core` keyword doesn't really make sense.  Perhaps we can catch if it's been set and print an error message?

**Comment by JonathonMisiewicz** [comment]
Your read of the paper sounds right to me. I think a better fix would be to temporarily unset `freeze_core` if it's been set, print a warning message, carry through with the computation, and then set it back. I'll wait to hear back from the author of the Psi implementation before making any changes, though.

**Comment by PeterKraus** [comment]
I'd also vote for catching and aborting with a descriptive error, as opposed to temporarily unsetting and carrying on. If the method is part-FC part-all electron, then FC shouldn't be touched in the input.

**Comment by JonathonMisiewicz** [comment]
The reason I stumbled across this error is because I included in the same input file as some computations with other, less totalitarian methods. This particular method doesn't even allow the user to change the basis set, so by this logic, it should crash whenever the user specifies a basis set, too.

My experience still has me in favor of run the computation but warn the user.

**Comment by PeterKraus** [comment]
>  This particular method doesn't even allow the user to change the basis set, so by this logic, it should crash whenever the user specifies a basis set, too.

This would be, to me, the expected behaviour, and should be applied consistently for all "composite" methods - including `pbeh-3c`, `hf-3c`. Because if you specify a basis different than def2-msvp for `pbeh-3c`, it's no longer `pbeh-3c`!

I'm happy to tackle this one once we agree what to do here.

**Comment by JonathonMisiewicz** [comment]
Let's start with the basics: Should it be possible to let a user do an SCF computation and then a G2 computation? I would think yes, but I can't see how that would be possible with your proposal. The obvious input is

```
molecule {
h
h 1 1.0
}

set basis sto-3g
energy('scf')
energy('gaussian-2')
```

If you don't set basis, SCF crashed. If you do, then under your proposal, the gaussian-2 would.

**Comment by PeterKraus** [comment]
It might not be in the code this way, but to me, the keyword `set basis something` sets a global-like value: everything that follows is expected to be calculated using that basis set. Both you and I know, that it is possible to change the code to allow:
```
set basis sto-3g
energy('gaussian-2')
```
to finish, and currently it even calculates a proper G2 energy. However, for example:
```
set basis sto-3g
energy('pbeh-3c')
```
doesn't - it will use `sto-3g` and it'll pretend it's doing `pbeh-3c`.

If I pretend I'm someone who's never used Psi4 before, this is confusing. I believe we should change it, one way or another, but be consistent for composite methods.

To answer your question, i'd suggest:
```
molecule {
...
}

energy('scf/sto-3g')
energy('gaussian-2')
```
as a neat, unambiguous alternative.

**Comment by JonathonMisiewicz** [comment]
I agree that `set basis X` should set a global. Are you saying that a method needing to locally override a global should disable that global being set in the first place?

**Comment by loriab** [comment]
General philosophy has been that something in the energy/grad/etc call can silently trump an option ( otherwise reference rhf + energy b3lyp wouldn’t work).  I think that holds for method, basis, fc, etc for g2. The hf3c isn’t so clear cut imo b/c the basis set is constant, so easy to switch out. Though as peter points out, it’s no longer hf3c then and shouldn’t get the “hf3c total energy “ set, only perhaps “scf total energy”. Comes down to does a method suggest or require certain options. 

At one point, fc in this g2 sequence were behaving very strangely for me.

**Comment by hokru** [comment]
I would expect that if I use a composite method keyword (G2, HF3c, etc.) that is will override anything silly I have set in the input or crash with a warning about inconsistent input. Way too many mistakes will be done otherwise.

**Comment by JonathonMisiewicz** [comment]
I'd also put G2 as _requiring_, so the method trumps the option.

I still advocate "override anything silly" over "crash with a warning". That one computation can't take certain user-set globals shouldn't disable setting the global for an earlier computation.

**Comment by PeterKraus** [comment]
I do have a feeling @loriab fixed this one, as the `FREEZE_CORE` option is stashed. In any case the MWE in the original post works with current master. Should this be closed?

**Comment by JonathonMisiewicz** [comment]
Confirming that the MWE works with master. For records purposes, can you link the PR that stashed the freeze core option? I'll close afterwards.

**Comment by PeterKraus** [comment]
It doesn't make sense, as the stashing was added 3 years ago. The change that fixed this must be related to some other `'FREEZE_CORE'` fix between 1.3 and current master.

**Comment by JonathonMisiewicz** [comment]
Figuring out which commit fixed this issue is now an exercise left to the programmer.
